mapie/control_risk/p_values.py

```diff
@@ -0,0 +1,119 @@
+from typing import Union, cast
+
+from mapie._typing import NDArray
+from mapie.utils import check_alpha
+
+import numpy as np
+from scipy.stats import binom
+
+
+def compute_hoeffdding_bentkus_p_value(
+    r_hat: NDArray,
+    n_obs: int,
+    alpha: Union[float, NDArray]
+) -> NDArray:
+    """
+    The method computes the p_values according to
+    the Hoeffding_Bentkus inequality for each
+    alpha.
+    We return the minimum between the Hoeffding and
+    Bentkus p-values (Note that it depends on
+    scipy.stats). The p_value is introduced in
+    learn then test paper [1].
+
+    Parameters
+    ----------
+    r_hat: NDArray of shape (n_lambdas, )
+        Empirical risk with respect
+        to the lambdas.
+        Here lambdas are thresholds that impact decision
+        making and therefore empirical risk.
+
+    n_obs: int.
+        Correspond to the number of observations in
+        dataset.
+
+    alpha: Union[float, Iterable[float]].
+        Contains the different alphas control level.
+        The empirical risk must be less than alpha.
+        If it is a iterable, it is a NDArray of shape
+        (n_alpha, ).
+
+    Returns
+    -------
+    hb_p_values: NDArray of shape (n_lambda, n_alpha).
+
+    References
+    ----------
+    [1] Angelopoulos, A. N., Bates, S., Candès, E. J., Jordan,
+    M. I., & Lei, L. (2021). Learn then test:
+    "Calibrating predictive algorithms to achieve risk control".
+    """
+    alpha_np = cast(NDArray, check_alpha(alpha))
+    alpha_np = alpha_np[:, np.newaxis]
+    r_hat_repeat = np.repeat(
+        np.expand_dims(r_hat, axis=1),
+        len(alpha_np),
+        axis=1
+    )
+    alpha_repeat = np.repeat(
+        alpha_np.reshape(1, -1),
+        len(r_hat),
+        axis=0
+    )
+    hoeffding_p_value = np.exp(
+        -n_obs * _h1(
+            np.where(
+                r_hat_repeat > alpha_repeat,
+                alpha_repeat,
+                r_hat_repeat
+            ),
+            alpha_repeat
+        )
+    )
+    bentkus_p_value = np.e * binom.cdf(
```

Comment:
are we 100% sure we need scipy here ? not very confortable with adding a new dependancy

Reply:
I'm not very comfortable with that either, but in practice, scipy is a dependency of scikit-learn. So there's no problem.